### PR TITLE
Decouple authenticator code in ci/test

### PR DESCRIPTION
**Pull request by jonahx**
### What does this PR do?

- _What's changed? Why were these changes made?_

This commit is a complete refactor of ci/test whose main purpose is to
decouple the "test runner" code from the individual tests.  This makes
the bash much easier to understand and maintain, and ensures that every
individual test is independent, and that its dependencies are clear.

This was accomplished by:

1. Stripping down "ci/test" to its bare essentials.  It now only
   contains 4 functions: main, help, finish (for cleanup work), and
   docker_diagnositics (for printing docker logs).

2. Creating a folder "ci/test_suites", whose subfolders are the
   individual tests.  The entrypoint for every test is simply the
   file named "test" in its folder.

3. Putting common functions shared by the tests into a "shared.sh" file,
   which can be sourced by the individual as needed.

4. Updating the "Jenkinsfile" to reference the tests using this new
   convention.

- _How should the reviewer approach this PR, especially if manual tests are required?_

One strategy would be:

1. Start by looking at the new `ci/test` file.
2. Next review`shared.sh`.
3. Finally look at the individual `test` files under `test_suites/**`

### What ticket does this PR close?

#1897 

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [x] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [x] This PR does not require updating any documentation

### CODE REVIEW

File: ci/test_suites/authenticators_oidc/test [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    conjur /authn-oidc/keycloak/scripts/fetchCertificate
+}
+
+# We also run an ldap-server container for testing the OIDC & LDAP combined use-case.
+# Note: We can't run the above use-case in a separate Jenkins step because we'll have a port bind for keycloak
+_prepare_env_auth_ldap
+_prepare_env_auth_oidc
+
+_run_cucumber_tests authenticators_oidc 'ldap-server oidc-keycloak' \
+  _hydrate_oidc_env_args
```
Comment (micahlee) [inline ci/test_suites/authenticators_oidc/test:117]: What are your thoughts on introducing a `main` function when you have other functions in the script? It's a personal preference, but I'd rather see the entrypoint flow at the top, with supporting functions beneath.

File: ci/test_suites/rotators/test [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+set -e
+
+# This is executed by the main "ci/test" script after cd-ing into "ci".
+# shellcheck disable=SC1091
+source "./shared.sh"
+
+_run_cucumber_tests rotators testdb
```
Comment (micahlee) [inline ci/test_suites/rotators/test:8]: Same as the other script, factoring `testdb` to a variable would make this more clear.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -306,6 +307,11 @@ $ ./cli exec
 root@9feae5e5e001:/src/conjur-server#
 ```
 
+### Adding New Test Suites
+
+When adding new test suites, please follow the guidelines in the top comments
+of the file `ci/test`.
```
Comment (micahlee) [inline CONTRIBUTING.md:313]: Could you make this a link to the file?

**Review by micahlee — COMMENTED** [review]
Nice work, @jonahx!

I left a few comments for your consideration, but overall this is a great improvement.

File: ci/test [hunk 4]
```diff
@@ -1,95 +1,50 @@
...
+# 3. Add an entry to the SUBCOMMANDS section of the "help" function below,
+#    describing your new test.
+#
+# Your test is now ready, and can be executed in the "Jenkinsfile" like:
+#
+#     sh 'ci/test my_new_test'
+#
+# IMPORTANT NOTES:
+#
+# 1. "test" is run from the "ci" folder.  So if your
```
[the remark this answers is not in the document]
Comment (jonahx) [reply ci/test:20]: I love this level of attention to detail, and don't mind changing, but what about the value of numbering for being able to refer to them?

File: ci/test_suites/authenticators_gcp/test [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    "GCE_SERVICE_ACCOUNT_EMAIL=${gce_vars[3]}"
+
+    "GCF_SERVICE_ACCOUNT_ID=${gcf_vars[0]}"
+    "GCF_SERVICE_ACCOUNT_EMAIL=${gcf_vars[1]}"
+  )
+}
+
+# Note: We pass the name of the function as the last arg, since we're
+# using namerefs.
+_run_cucumber_tests authenticators_gcp "" _hydrate_gcp_env_args
```
[the remark this answers is not in the document]
Comment (jonahx) [reply ci/test_suites/authenticators_gcp/test:63]: Sure I'll add a comment.  If you have a better idea lmk, but we are limited by bash not supporting named args 😞

File: ci/test_suites/authenticators_ldap/test [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+set -e
+
+# This is executed by the main "ci/test" script after cd-ing into "ci".
+# shellcheck disable=SC1091
+source "./shared.sh"
+
+_prepare_env_auth_ldap
```
[the remark this answers is not in the document]
Comment (jonahx) [reply ci/test_suites/authenticators_ldap/test:8]: This was a carryover name, which you've correctly identified as being too vague.  I'll change it to `start_ldap_server`, which is more accurate.

File: ci/test_suites/authenticators_ldap/test [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+_run_cucumber_tests authenticators_ldap ldap-server
```
[the remark this answers is not in the document]
Comment (jonahx) [reply ci/test_suites/authenticators_ldap/test:9]: I made this change but note: since we're not in a function we can't use "local".  Instead, this file is _executed_, which means that it runs in a subshell, which means it's variables are protected from leaking that way.

File: ci/docker-compose.yml [hunk 8]
```diff
@@ -55,7 +55,8 @@ services:
     command: server
     volumes:
       - authn-local:/run/authn-local
-      - ./authn-oidc/keycloak:/authn-oidc/keycloak/scripts
+      # TODO: authenticators_oidc/test has a dep on this
```
Comment (sigalsax) [inline ci/docker-compose.yml:58]: is this a TODO that will be addressed in this PR?

File: ci/test_suites/authenticators_gcp/get_gcp_id_tokens.sh [hunk 9]
```diff
@@ -122,18 +127,20 @@ get_tokens_into_files() {
...
 
-  local tokens="$(cat $tokens_config_file)"
+  local tokens
+  tokens=$(< "$tokens_config_file")
   local token_prefix="${TOKENS_OUT_DIR_PATH}/${TOKEN_FILE_NAME_PREFIX}"
 
   for row in $(echo "${tokens}" | jq -c '.[]'); do
     _jq() {
-      echo ${row} | jq -r ${1}
+      echo "$row" | jq -r "$1"
```
Comment (sigalsax) [inline ci/test_suites/authenticators_gcp/get_gcp_id_tokens.sh:143]: Ive seen both `${}` and just `$`. Are they syntactically the same? Can we choose one align throughout the repo?

File: ci/test (hunk 4 above)
[the remark this answers is not in the document]
Comment (micahlee) [reply ci/test:20]: This was just something I picked up from docs reviews with the technical writers. And even for reference, numbers aren't a good choice for labels, unless the order itself is meaningful. If reference is important, it's probably better to factor these notes out to a `.md` file with headings that can be linked to.

Are we referring to them somewhere now?

That said, this is not a blocking comment, I don't feel that strongly either way. Just passing on what's been shared with me.

File: ci/test_suites/authenticators_ldap/test (hunk 6 above)
[the remark this answers is not in the document]
Comment (micahlee) [reply ci/test_suites/authenticators_ldap/test:8]: Ah, `start_ldap_server` makes it much more obvious. I first thought this had to do with environment variables.

File: ci/test_suites/authenticators_gcp/test (hunk 5 above)
[the remark this answers is not in the document]
Comment (micahlee) [reply ci/test_suites/authenticators_gcp/test:63]: Yeah, the only other idea I had was factoring out the variable even for an empty argument so that it's clear that the argument is for `additional_services`, but that this test just doesn't require any.

Again, this is a nit level comment. Just trying to make it easier to do the right thing if coming to add a test suite with zero prior experience with this (or having not touched it in a while). :)

File: ci/test_suites/authenticators_oidc/test (hunk 1 above)
Comment (jonahx) [reply ci/test_suites/authenticators_oidc/test:117]: I think it's a good pattern, I'll make that change.

File: ci/docker-compose.yml (hunk 8 above)
Comment (jonahx) [reply ci/docker-compose.yml:58]: I will go thru and pull them out for remaining work.

File: ci/test_suites/authenticators_gcp/get_gcp_id_tokens.sh (hunk 9 above)
Comment (jonahx) [reply ci/test_suites/authenticators_gcp/get_gcp_id_tokens.sh:143]: @sigalsax You should use `$1` or `$var` when you can, and `${1}` or `${var}` when you have to.  In some contexts they are the same, in others not, so you can't unilaterally say to always use one.  Rather, you _could_ decide to always use `${var}` but you'll add a lot of visual noise for no reason to the 90% case.  Better to understand what they actual do.

When you are using arrays or parameter expansions, the `${}` form is required.  It's also required if you want to interpolate your variable and have text directly abutting it: 

```sh
$ sigal=avocado
$ echo "$sigals RULE"
 RULE
$ echo "${sigal}s RULE"
avocados RULE
```
There are probably some other cases where you need `${}` I'm forgetting too.  But most of the time, for simple cases, the plain form `$var` works fine and should be preferred.

File: ci/test_suites/authenticators_gcp/test (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonahx) [reply ci/test_suites/authenticators_gcp/test]: I implemeted the `additional_services` idea.

File: ci/test_suites/authenticators_gcp/get_gcp_id_tokens.sh (hunk 9 above)
Comment (sigalsax) [reply ci/test_suites/authenticators_gcp/get_gcp_id_tokens.sh:143]: Gotcha and might I say, love that I am equated to an 🥑

**Review by micahlee — DISMISSED** [review]
LGTM! Let me know when you rebase and I can re-approve.

**Review by micahlee — APPROVED** [review]
